packages/api/pages/api/languages/[code]/verses/[verseId]/glosses.ts

```diff
@@ -160,4 +167,113 @@ export default createRoute<{ code: string; verseId: string }>()
+    schema: z.object({
+      data: z.record(
+        z.string(),
+        z.object({
+          gloss: z.string().optional(),
+          state: z
+            .enum(Object.values(GlossState) as [GlossState, ...GlossState[]])
+            .optional(),
+        })
+      ),
+    }),
+    authorize: authorize((req) => ({
+      action: 'translate',
+      subject: 'Language',
+      subjectId: req.query.code,
+    })),
+    async handler(req, res) {
+      const language = await client.language.findUnique({
+        where: { code: req.query.code },
+      });
+      if (!language) {
+        return res.notFound();
+      }
+
+      const glosses = Object.fromEntries(
+        (
+          await client.$queryRaw<PrismaTypes.Gloss[]>`
+        SELECT "wordId", "gloss", "state" FROM "Gloss" JOIN "Word" ON "Gloss"."wordId" = "Word"."id"
+          WHERE "Word"."verseId" = ${req.query.verseId} AND "Gloss"."languageId" = ${language.id}::uuid
+      `
+        ).map(({ wordId, ...data }) => [wordId, data])
+      );
+
+      const entriesToPatch = Object.entries(req.body.data)
+        .map(([wordId, { gloss, state }]) => ({
+          wordId,
+          gloss,
+          // If the gloss is empty, the state should be unapproved
+          state: gloss === '' ? GlossState.Unapproved : state,
+        }))
+        .filter(
+          // If the gloss is undefined and the state is undefined, nothing will be updated, making the entry useless.
+          ({ gloss, state }) => gloss !== undefined || state !== undefined
+        );
+
+      const entriesToInsert = entriesToPatch.filter(
+        ({ wordId }) => !glosses[wordId]
+      );
+      const entriesToUpdate = entriesToPatch.filter(({ wordId, ...data }) => {
+        const oldGloss = glosses[wordId];
+        return (
+          !!oldGloss &&
+          // Only include updates where something has actually changed.
+          ((data.gloss && data.gloss !== oldGloss.gloss) ||
+            (data.state && data.state !== oldGloss.state))
+        );
+      });
+
+      await client.gloss.createMany({
+        data: entriesToInsert.map((data) => ({
+          languageId: language.id,
+          ...data,
+        })),
+      });
+      await client.glossHistoryEntry.createMany({
+        data: entriesToInsert.map((data) => {
+          return {
+            languageId: language.id,
+            ...data,
+            userId: req.session?.user?.id,
+            source: GlossSource.User,
+          };
+        }),
+      });
+
+      if (entriesToUpdate.length > 0) {
+        await client.$executeRaw`
+          UPDATE "Gloss" SET 
+            "gloss" = COALESCE(gloss_updates."gloss", "Gloss"."gloss"),
+            "state" = COALESCE(gloss_updates."state", "Gloss"."state")
+          FROM (VALUES ${Prisma.join(
+            entriesToUpdate.map(
+              ({ wordId, gloss, state }) =>
+                Prisma.sql`(${wordId}, ${gloss ?? null}, ${
+                  state ?? null
```

Comment:
Would it be possible to skip using a transaction and instead use `WITH` clauses?

Reply:
Why is it better to skip a transaction? We can just do this:
```ts
await client.$transaction(async (tx) => {
  // Use tx instead of client to run the queries, logic, and upserts
})
```

Reply:
Done.

Reply:
Done.